webrender/src/frame_builder.rs

```diff
@@ -1684,96 +1678,54 @@ impl FrameBuilder {
             .get(&pipeline_id)
             .expect("No display list?")
             .display_list;
-        debug!(
-            "\tclip_bounds {:?}, layer_local_clip {:?}",
-            clip_bounds,
-            packed_layer.local_clip_rect
-        );
 
         if !stacking_context.is_backface_visible && packed_layer.transform.is_backface_visible() {
             return;
         }
 
+        let prim_context = PrimitiveContext::new(
+            packed_layer_index,
+            packed_layer,
+            clip_and_scroll.clip_node_id(),
+            screen_rect,
+            clip_scroll_tree,
+            &self.clip_store,
+            device_pixel_ratio,
+        );
+
+        let prim_context = match prim_context {
+            Some(prim_context) => prim_context,
+            None => return,
+        };
+
+        debug!(
+            "\tclip_bounds {:?}, layer_local_clip {:?}",
+            prim_context.clip_bounds,
+            packed_layer.local_clip_rect
+        );
+
         for i in 0 .. prim_count {
             let prim_index = PrimitiveIndex(base_prim_index.0 + i);
-            let (prim_local_rect, prim_screen_rect) = match self.prim_store.build_bounding_rect(
-                prim_index,
-                &clip_bounds,
-                &packed_layer.transform,
-                &packed_layer.local_clip_rect,
-                device_pixel_ratio,
-            ) {
-                Some(rects) => rects,
-                None => continue,
-            };
-
-            debug!("\t\t{:?} bound is {:?}", prim_index, prim_screen_rect);
 
-            let prim_metadata = self.prim_store.prepare_prim_for_render(
+            if let Some((prim_local_rect, prim_screen_rect)) = self.prim_store.prepare_prim_for_render(
```

Comment:
note that we have a special type `clip::Geometry` for having both the local and global rectangles